### ISSUE TITLE
Query with time < now() produces negative timestamps and null data

### ISSUE THREAD

**Issue by xPaw** [influxdata/influxdb#8690]
InfluxDB 1.3.2 on Debian 9.

This query worked as I expect it to (only return existing data and fill between first and last points) in 1.2.x, but now it starts with completely odd negative timestamp (like -9223372036) which breaks my stuff entirely.

Was this an expected change, as I don't see anything particular in the changelog.

The full query is: `SELECT time, MAX(value) FROM concurrent_players_daily WHERE appid = '440' AND time < now() - 1d GROUP BY time(1d) fill(linear)`

This appears to directly related to `GROUP BY` clause as removing `fill(linear)` does not change anything.

**Comment by jsternberg** [comment]
We added the ability to query from times before 1970 so the lower limit of a query is now approximately -9223372036. I don't know if that's the exact number, but I believe it's close to there.

You must specify a lower limit. There is a bug fix in 1.4 that will throw an error if you don't specify a lower limit on this type of query though to prevent this kind of error.

**Comment by xPaw** [comment]
@jsternberg So what's the approach running a query that would return all the stored data? The query without a lower bound worked before upgrading to 1.3.

**Comment by jsternberg** [comment]
I'm not sure what you mean, but the lower timestamp has been that way since 1.0. The issue documenting the fix is in #7212.

If you want something that aggregates all data, you can use `fill(none)` and that will not generate the extra intervals when there are no points.

It also appears like I may be wrong about the lower limit in 1.4. I was going to add that limitation to prevent overflowing queries, but it seems like I never did and I'll avoid adding that limitation so that this kind of query can work. But, keep an eye out for the changelog on this because it's very easy for someone to forget to limit their query and create a slow query if they don't specify `fill(none)`.

**Comment by xPaw** [comment]
This definitely worked on 1.2 as I used the exact same query which started returning negative timestamps on 1.3 with the exact same query, so something must have changed in that regard.

**Comment by jsternberg** [comment]
I'm curious why this was working for you in 1.2. Can you give me an example with some sample data so I can confirm this isn't a regression? The behavior I discussed should have existed since 1.0.

**Comment by xPaw** [comment]
The query I used is in the first post. Data would be just couple of numbers for couple of days. I don't think the data even matters.

**Comment by xPaw** [comment]
@jsternberg So what would be a solution to select all existing data and grouping it without knowing the specific timestamp where the data starts?

I can't do something like this either because influxdb does not allow selecting just time:
```
time > (SELECT first(value) FROM concurrent_players_daily WHERE appid = '440')
```

**Comment by rbetts** [comment]
@xPaw if you can change `fill(linear)` to `fill(none)`, that would skip empty intervals. However it also changes the behavior for missing intervals in the more recent data. 

@jsternberg Any luck reproducing this?

**Comment by rbetts** [comment]
@mark-rushakoff - can we make sure we have this query pattern in h/t?

**Comment by jsternberg** [comment]
I bisected it back to timezone support: 347b018.

I'll start looking into it.

**Comment by jsternberg** [comment]
So... this gets a little trippy. There are two bugs. There are two bugs in 1.2 and one bug in 1.3. The two bugs, in combination with each other, caused the behavior that @xPaw saw. That behavior is incorrect in the current query engine.

Before I continue, I actually see @xPaw's concern and I think we should consider ways to make his behavior be correct. I've got an idea, but I'd prefer to focus on fixing the bugs first.

Here's the fixed bug:

```diff
        if itr.opt.Ascending {
-               itr.window.time = p.Time + int64(itr.opt.Interval.Duration)
+               itr.window.time += int64(itr.opt.Interval.Duration)
        } else {
-               itr.window.time = p.Time - int64(itr.opt.Interval.Duration)
+               itr.window.time -= int64(itr.opt.Interval.Duration)
+       }
```

So the timezone feature fixed one of the two bugs because I was running around that section of the code. The next interval for a fill iterator is supposed to be based on the last interval. Both sections of code should work the same when there are no other bugs.

But, I discovered a second bug. The initial timestamp of the interval window is being calculated incorrectly with `Window()`. When the time is the `MinTime`, this underflows and resets to the highest time. The fill iterator has a feature where if points are interspered before the expected next window, it will output them. So the fill iterator is calculating, due to underflow, that the window should be completely empty, but the fill iterator is still letting other points through because of that other feature.

We're going to need to fix the underflow problem first and then I think we can make it so when the fill iterator sees the min time, it calculates the time from the first point in the interval instead of using the start time. We use that in some other areas too so we already have precedent for doing that.

**Comment by jsternberg** [comment]
I appear to have made a mistake while looking at the 1.2 code. There were two bugs, but the underflow bug was present in 1.2 and was fixed later so it doesn't exist in 1.3.

I'll start looking into a way that we can make a non-existent lower bound to be the most recent window. About a month ago, I merged a change that made it an error not to specify a lower bound to avoid this kind of issue.

Sorry about that. I actually forgot I made that change and didn't realize it until I tried the query on the master branch today. I'll try to turn the 1.2 behavior from what was a bug into something that is a feature.
